src/code/snap_operator.py

```diff
@@ -54,6 +56,170 @@ def find_target(config: configs.Config, suffix: str) -> Optional[snap_holder.Sna
+            columns.append(f"  {snap_timestamp}")
+
+            trigger_str = "".join(
+                c if snapshot.metadata.trigger == c else " " for c in "SIU"
+            )
+            columns.append(trigger_str)
+
+            elapsed = (now - snapshot.snaptime).total_seconds()
+            elapsed_str = f"({human_interval.humanize(elapsed)} ago)"
+            columns.append(f"{elapsed_str:<20}")
+            columns.append(snapshot.metadata.comment)
+
+            print("  ".join(columns))
+
+        print()
+
+    confirm = input("Are you sure you want to delete the above snapshots?  [y/N]")
+    match confirm:
+        case "y" | "Y" | "yes" | "Yes" | "YES":
+            return True
+        case _:
+            return False
+
+
+def _indicator_filter(
+    snapshots_of_configs: dict[str, dict[datetime.datetime, snap_holder.Snapshot]],
+    indicator: str,
+) -> dict[str, dict[datetime.datetime, snap_holder.Snapshot]]:
+    return {
+        config_name: {
+            datetime_: snapshot
+            for datetime_, snapshot in datetime_and_snapshots.items()
+            if snapshot.metadata.trigger == indicator.upper()
+        }
+        for config_name, datetime_and_snapshots in snapshots_of_configs.items()
+    }
+
+
+def _snapshots_of_configs(
+    configs_iter: Iterable[configs.Config],
+) -> dict[str, dict[datetime.datetime, snap_holder.Snapshot]]:
+    """Find the snapshots held by each configuration file."""
+    mapping: dict[str, dict[datetime.datetime, snap_holder.Snapshot]] = {}
+
+    # Create a mapping between configuration files and snapshots.
+    for config in configs_iter:
+        config_name = Path(config.config_file).stem
+        mapping[config_name] = {}
+        # Create a mapping between snapshot timestamps and snapshots
+        # to facilitate later filtering and sorting of the snapshots.
+        for snapshot in _get_old_backups(config):
+            mapping[config_name][snapshot.snaptime] = snapshot
+
+    return mapping
+
+
+def _convert_to_timestamp(suffix: str) -> str:
+    """Convert human-readable time format into a timestamp."""
+    if not _human_friendly_timestamp(suffix):
+        return suffix
+
+    try:
+        # "2024-11-19_15:30".split("_") -> ["2024-11-19", "15:30"]
+        date, time = suffix.split("_")
+    except ValueError:  # unpack error
+        date = suffix
+        # If the user does not provide the hour and minute, it defaults to 23:59
+        hour_minute = "2359"
+    else:
+        # "15:30".split(":") -> ["15", "30"] -> "1530"
+        # "1530".split(":") -> ["1530"] -> "1530"
+        hour_minute = "".join(time.split(":"))
+        suffix_not_minutes = len(hour_minute) == 2
+        if suffix_not_minutes:
+            hour_minute += "00"
+
+    days_not_padded_zeros = r"^\d{4}-\d{2}-\d$"
+    pad_days_with_zeros = r"^\d{4}-\d{2}-0\d$"
+    year_month_day = re.sub(days_not_padded_zeros, pad_days_with_zeros, date)
+    return year_month_day + hour_minute
```

Comment:
## Premise

Assume there is 1 configuration and 5 snapshots:

```txt
Config: xxx/y.conf
Snaps:
  1 202412131055    U
  2 202412132359  S

  3 202412140810    U
  4 202412142359  S

  5 202412151756    U
```

Using `--start TIMESTAMP`, `--end TIMESTAMP`, and `--start TIMESTAMP1 --end TIMESTAMP2`:

1. It should delete all snapshots, but actually deletes snapshot number `2` and all subsequent snapshots.
    - `yabsnap batch-delete --start 2024-12-13` (_23:59)\
    The ideal timestamp completion should be:
    - `yabsnap batch-delete --start 2024-12-13` (_00:00)
2. It should delete snapshots before number `4`, and it actually deletes snapshots before number `4` (the timestamp completion here is correct).
    - `yabsnap batch-delete --end 2024-12-14` (_23:59)
3. It should delete snapshots `1-3`, but actually deletes `2` and `3`.
    - `yabsnap batch-delete --start 2024-12-13 --end 2024-12-14`\
    This is equivalent to:
    - `yabsnap batch-delete --start 2024-12-13_23:59 --end 2024-12-14_23:59`\
    The ideal completion for the above shorthand dates should be:
    - `yabsnap batch-delete --start 2024-12-13_00:00 --end 2024-12-14_23:59`

## Issue

After reviewing the above examples, I believe you can understand that the current `_convert_to_timestamp` function has flaws.

It has the following issues:

- When the timestamp provided by the user has a precision of at most days (e.g. `2024-01-01`), the timestamp completion is inaccurate.
- The design for zero-padding of months and days, as well as the completion of minutes, is not well-implemented (the following are simplified examples):
  - `2024-5-01` -> `2024-05-01`
  - `2024-01-5` -> `2024-01-05`
  - `2024-01-01_02` -> `2024-01-01_02:00`

## Thoughts

When writing the `_convert_to_timestamp` function, I did not consider the introduction of state. When the timestamp provided by the user has a precision of at most days, it is necessary to know which command-line optional parameters (`--start`, `--end`) the user has used so that the `_convert_to_timestamp` function can provide accurate timestamp completion.

Currently, regardless of which optional parameter (`--start`, `--end`) is used, the timestamp completion is always `_23:59`, which is unreasonable. However, introducing state would increase the complexity of the code, which exceeds my expectations.

My thought is to abandon the implementation of "**diverse** human-friendly timestamp parameters" and instead limit users to only input specific formats of timestamps, raising the upper limit to reduce the complexity of functionality implementation and avoid unintentional invalid input from users. What are your thoughts and suggestions on this?

Reply:
Then I will follow my own idea to implement it: the human-friendly timestamp format will only support the ISO 8601 standard.

Reply:
> 100% agree. We can choose to make it more complicated later if we need to.

😂

> As for the format we demand from the users, WDYT of using the same format as is shown in "yabsnap list", i.e. YYYYMMDDhhmmss, e.g. --start=202412131055 --end=202412142359? It will keep it consistent, while avoiding confusion.

In some cases, it's difficult for me to visually separate the numbers that are stuck together and associate them with dates, times, etc. So, I still kept the option "allow input in ISO 8601 date format."

> Another thought - if not now then later, I feel we should have a _is_within_timescope(to_checl:datetime, start: str, end: str). This is the logical unit that find_multi_targets() can directly rely on. Also this is the intricate part, and this is the function that can/should be unit-tested (e.g. with similar examples as you have stated in your comment).

I think [TimeScopeFilter](https://github.com/Tql-ws1/yabsnap/commit/f0eb2a69faa94900637ccaf1a5a8465b21f020a4#diff-aca5df2f16d84600a950ca86b8f64bdf8c61ef6ff0809ba813736c1a7af61bd7R58) has already achieved this 👀